api/agent/call.go

```diff
@@ -368,28 +338,82 @@ func (a *agent) GetCall(opts ...CallOpt) (Call, error) {
 	return &c, nil
 }
 
-func setupCtx(c *call) {
-	ctx, _ := common.LoggerWithFields(c.req.Context(),
-		logrus.Fields{"id": c.ID, "app_id": c.AppID, "route": c.Path})
-	c.req = c.req.WithContext(ctx)
+// setCallPayload sets the GetBody method on the call, if one doesn't exist
+func (c *call) setCallPayload(ctx context.Context, input io.Reader, getBody func() (io.ReadCloser, error)) error {
+	if getBody != nil {
+		c.getBody = getBody
+		return nil
+	}
+
+	// if it has a string method, shortcut that (say, a bytes buffer...)
+	// TODO we could check Bytes() as well but this whole thing should die anyway, hedging...
+	if stringer, ok := input.(interface{ String() string }); ok {
+		c.getBody = func() (io.ReadCloser, error) {
+			return ioutil.NopCloser(strings.NewReader(stringer.String())), nil
+		}
+		return nil
+	}
+
+	// only allow one read of the buffer ever. everybody else must wait. this is
+	// entirely fucked up and can deadlock all succeeding threads for this body,
+	// but go ahead and try just reading the reader into a buffer and see if life
+	// is still worth living. TODO this leaves a situation where if the first reader
+	// doesn't finish then nobody else gets the body ever. NO SOUP FOR YOU
+	done := make(chan struct{}, 1)
+	done <- struct{}{} // prime it
+
+	buf := bufPool.Get().(*bytes.Buffer)
```

Comment:
I think this leaks. Should there be a corresponding Put()?

Reply:
can add one to the returned io.Closer `Close()` instead of the nop. d'oh. it won't leak, per se, it'll get GC'd but we're stealing from the pool without giving back (and probably not allocating new ones only while throughput is decreasing).